src/factor.cc

```diff
@@ -23,17 +23,37 @@ Factor::Factor(factor_id_t id, feature_value_t feature_value,
       weight_id(weight_id),
       func_id(func_id),
       n_variables(n_variables),
-      n_start_i_vif(INVALID_ID),
-      factor_params(NULL) {}
+      n_start_i_vif(INVALID_ID) {}
 
 Factor::Factor(const Factor &rf)
     : id(rf.id),
       feature_value(rf.feature_value),
       weight_id(rf.weight_id),
       func_id(rf.func_id),
       n_variables(rf.n_variables),
-      n_start_i_vif(rf.n_start_i_vif),
-      factor_params(rf.factor_params) {}
+      n_start_i_vif(rf.n_start_i_vif) {
+  factor_params.reset(
+      rf.factor_params
+          ? new std::unordered_map<factor_weight_key_t, FactorParams>(
+                *rf.factor_params)
+          : nullptr);
```

Comment:
Lets keep this also in the initialization list. Or we could just use the operator= to define the copy constructor (`*this = other`) as in Variable, which dedups the code but I guess in the opposite way. The *unifying assignment operator* is the right way to go, but let's do that later for all classes. https://en.m.wikibooks.org/wiki/More_C%2B%2B_Idioms/Copy-and-swap 
Having the copy constructor first call the default may be a safe thing to do.